### PR TITLE
Data source for aws_service_discovery_dns_namespace

### LINKED ISSUE #6384: Data source for aws_service_discovery_private_namespace
_This issue was originally opened by @noobingitup as hashicorp/terraform#19306. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### 0.11.1

I tried looking through the terraform documentation for namespace data sources and could not find any.

The use case for a namespace data source is instead of passing a namespace id and the respected hosted zone to a module, one could use pass the namespace name and data source the namespace for the id and hosted zone id making the module more readable and understandable when peering into the terraform code.

### PR DESCRIPTION
**Pull request by saravanan30erd**
Fixes #6384 

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSServiceDiscoveryDnsNamespace_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSServiceDiscoveryDnsNamespace_basic -timeout 120m
=== RUN   TestAccAWSServiceDiscoveryDnsNamespace_basic
=== PAUSE TestAccAWSServiceDiscoveryDnsNamespace_basic
=== CONT  TestAccAWSServiceDiscoveryDnsNamespace_basic
--- PASS: TestAccAWSServiceDiscoveryDnsNamespace_basic (175.88s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws
...
```

### CODE REVIEW

File: aws/data_source_aws_service_discovery_dns_namespace.go
Comment (mbfrahry) [inline aws/data_source_aws_service_discovery_dns_namespace.go]: We should probably use the constants in the aws sdk in case these ever change in the sdk. You can find those here https://github.com/aws/aws-sdk-go/blob/master/service/servicediscovery/api.go#L6076

Comment (saravanan30erd) [reply aws/data_source_aws_service_discovery_dns_namespace.go]: @mbfrahry  thanks, its done.

$ make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic -timeout 120m
=== RUN   TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic
=== PAUSE TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic
=== CONT  TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic
--- PASS: TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic (168.80s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	168.848s

**Comment by saravanan30erd** [comment]
$ make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic -timeout 120m
=== RUN   TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic
=== PAUSE TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic
=== CONT  TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic
--- PASS: TestAccDataSourceAWSServiceDiscoveryDnsNamespace_basic (168.80s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	168.848s

**Comment by borgstrom** [comment]
@mbfrahry -- Looks like this one is ready to go.  Could you give it another review please?  This data source will make my life so much easier!  :)

**Comment by saravanan30erd** [comment]
@bflad  Can we push this? anything I need to do from my end?

**Comment by deedoz** [comment]
Hello,
Good job !
I miss this datasource also, looking forward for the merge.

**Comment by charlyx** [comment]
Great work @saravanan30erd 👏(I believe there is now a button to request a new review).

Hi @mbfrahry  👋
Thanks for your review. It looks like the changes you've requested have been done.
Is there anything else blocking the merge of this PR? I'm really looking forward to using it 🙏

**Comment by imranjavaid** [comment]
When can we expect a merge of this?

**Comment by loivis** [comment]
@mbfrahry @bflad merge?

**Comment by completenovice** [comment]
Hello, what is the status of this, please? Is this still being developed on?

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Action Required: Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by ewbankkit** [comment]
@saravanan30erd Thanks for the contribution and apologies for the delay in responding.
Are you able to address the changes mentioned in the comment above?

File: aws/data_source_aws_service_discovery_dns_namespace.go
Comment (ewbankkit) [inline aws/data_source_aws_service_discovery_dns_namespace.go:20]: I suggest renaming this to simply `type` to match the AWS API.

**Comment by saravanan30erd** [comment]
@ewbankkit  sure, I will rename the attribute and update you.

**Comment by ewbankkit** [comment]
Related:

* https://github.com/terraform-providers/terraform-provider-aws/issues/14881

File: aws/data_source_aws_service_discovery_dns_namespace.go
Comment (saravanan30erd) [reply aws/data_source_aws_service_discovery_dns_namespace.go]: @ewbankkit  Done

**Comment by saravanan30erd** [comment]
@ewbankkit  I renamed the attr `dns_type` to `type`. Please check.

**Comment by dayglojesus** [comment]
Original submission on this was December 2018?? holy cow. Talk about contributor purgatory. I cannot believe the author stuck with it as long as they did. ping @ewbankkit 

Is there no way for this to see daylight?

**Comment by saravanan30erd** [comment]
@anGie44 please check this PR

**Comment by anGie44** [comment]
Output of acceptance tests (Commercial):
```
--- PASS: TestAccAWSServiceDiscoveryDnsNamespaceDataSource_public (106.18s)
--- PASS: TestAccAWSServiceDiscoveryDnsNamespaceDataSource_private (114.47s)
```

Output of acceptance tests (Gov Cloud):
```
--- SKIP: TestAccAWSServiceDiscoveryDnsNamespaceDataSource_public (22.31s)
--- SKIP: TestAccAWSServiceDiscoveryDnsNamespaceDataSource_private (25.67s)
```

**Review by anGie44 — APPROVED** [review]
Hi again @saravanan30erd , thank you for this PR! Overall looked great 👍 Major thing was getting the PR aligned with new conventions used in the provider as we've migrated over to the V2 of the terraform plugin SDK. Plan to get this into the upcoming release later this week.

**Comment by anGie44** [comment]
Output of acceptance tests:
```
--- PASS: TestAccAWSServiceDiscoveryPublicDnsNamespace_disappears (359.05s)
--- PASS: TestAccAWSServiceDiscoveryPublicDnsNamespace_basic (369.10s)
--- PASS: TestAccAWSServiceDiscoveryDnsNamespaceDataSource_public (400.97s)
--- PASS: TestAccAWSServiceDiscoveryHttpNamespace_disappears (401.87s)
--- PASS: TestAccAWSServiceDiscoveryPrivateDnsNamespace_error_Overlap (433.00s)
--- PASS: TestAccAWSServiceDiscoveryPublicDnsNamespace_Description (455.22s)
--- PASS: TestAccAWSServiceDiscoveryService_disappears (508.90s)
--- PASS: TestAccAWSServiceDiscoveryHttpNamespace_Description (509.43s)
--- PASS: TestAccAWSServiceDiscoveryDnsNamespaceDataSource_private (510.19s)
--- PASS: TestAccAWSServiceDiscoveryHttpNamespace_basic (510.77s)
--- PASS: TestAccAWSServiceDiscoveryPrivateDnsNamespace_disappears (511.99s)
--- PASS: TestAccAWSServiceDiscoveryService_http (557.40s)
--- PASS: TestAccAWSServiceDiscoveryPrivateDnsNamespace_Description (558.17s)
--- PASS: TestAccAWSServiceDiscoveryPublicDnsNamespace_Tags (601.90s)
--- PASS: TestAccAWSServiceDiscoveryPrivateDnsNamespace_basic (602.43s)
--- PASS: TestAccAWSServiceDiscoveryService_Tags (674.79s)
--- PASS: TestAccAWSServiceDiscoveryHttpNamespace_Tags (675.31s)
--- PASS: TestAccAWSServiceDiscoveryService_public (675.98s)
--- PASS: TestAccAWSServiceDiscoveryService_private (704.14s)
--- PASS: TestAccAWSServiceDiscoveryPrivateDnsNamespace_Tags (712.03s)
```

**Comment by dayglojesus** [comment]
huzzah!

**Comment by oscargarciajr** [comment]
Do we have documentation on how to use this?

data "aws_service_discovery_dns_namespace" "controller" {
  type = "type"
  namespace = "namespace"
  description = "Serverless discovery managed zone."
}

**Pull request merged**
